csrc/id_model/loop_promotion.cpp

```diff
@@ -898,4 +932,60 @@ std::unordered_map<ValGroup, IterDomain*> LoopPromotionMapBuilder::get(
   return builder.build();
 }
 
+std::unordered_map<ValGroup, IterDomain*> LoopPromotionMapBuilder::
+    buildWithNoBroadcast() {
+  const auto& loop_graph = idGraph(IdMappingMode::LOOP);
+
+  std::unordered_map<ValGroup, IterDomain*> map;
+  for (const ValGroup& loop_group :
+       loop_graph.disjointValSets().disjointSets()) {
+    NVF_ERROR(!loop_group->empty());
+
+    // Any domain of this loop group can be the promotion ID. Try to
+    // find the simplest one, which means:
+    //
+    // - Prefer IDs with a constant extent
+    // - Prefer IDs with an extent that consists of a smaller number
+    // of exprs.
+
+    IterDomain* promotion = nullptr;
+    int64_t num_exprs = 0;
+    bool is_const = false;
+
+    for (Val* val : *loop_group) {
+      IterDomain* loop_id = val->as<IterDomain>();
+      auto this_num_exprs =
+          (int64_t)StmtSort::getExprsTo({loop_id->extent()}).size();
+      auto this_is_const = loop_id->extent()->isConstInt();
+
+      // First ID
+      if (promotion == nullptr) {
+        is_const = this_is_const;
+        promotion = loop_id;
+        num_exprs = this_num_exprs;
+        continue;
+      }
+
+      // If new ID is non-const while the current promotion is const,
+      // or if both IDs are const or non-const and the number of
+      // expressions is not smaller, keep the current promotion
+      if ((is_const && !this_is_const) ||
+          (is_const == this_is_const && this_num_exprs >= num_exprs)) {
+        continue;
+      }
+
+      // Update the current promotion
+      is_const = this_is_const;
+      promotion = loop_id;
+      num_exprs = this_num_exprs;
+    }
```

Comment:
This part is completely optional. Just make it easier to my eyes when looking at debug print and also it avoids this test error. https://github.com/NVIDIA/Fuser/pull/3079#issuecomment-2387571701